tools/omero/omero_import.xml

```diff
@@ -0,0 +1,125 @@
+<tool id="omero_import" name="OMERO import" version="@TOOL_VERSION@+galaxy@VERSION_SUFFIX@" profile="20.01" license="MIT">
+    <description></description>
+    <macros>
+        <token name="@TOOL_VERSION@">5.18.0</token>
+        <token name="@VERSION_SUFFIX@">0</token>
+    </macros>
+    <xrefs>
+        <xref type="bio.tools">omero</xref>
+    </xrefs>
+    <requirements>
+        <requirement type="package" version="@TOOL_VERSION@">omero-py</requirement>
+        <requirement type="package" version="21.0.2">openjdk</requirement>
+    </requirements>
+    <command detect_errors="exit_code"><![CDATA[
+        mkdir folder &&
+        #for $image in $folder:
+            ln -s '$image' folder/'$image.element_identifier'.$image.ext &&
+        #end for
+
+        omero import folder
+
+        #if $predefined_cond.predefined_sel == 'yes'
+            -s $predefined_cond.omero_host.fields.server
+            -p $predefined_cond.omero_host.fields.port
+            -u $__user__.extra_preferences.get('omero_'+$predefined_cond.omero_host.fields.value+'|username', $test_username)
+        #else
+            -s $predefined_cond.omero_host
+            -p $predefined_cond.omero_port
+            -u $__user__.extra_preferences.get('omero|username', $test_username)
+        #end if
+        > $log
+    ]]></command>
+    <environment_variables>
+        <environment_variable name="OMERO_PASSWORD"><![CDATA[#if $predefined_cond.predefined_sel == 'yes':
+$__user__.extra_preferences.get('omero_'+$predefined_cond.omero_host.fields.value+'|password', $test_password)
+#else
+$__user__.extra_preferences.get('omero|password', $test_password)
+#end if]]></environment_variable>
+    </environment_variables> 
+    <inputs>
+        <param name="folder" type="data" format="bmp" multiple="true" optional="false" label="Images"/>
+        <conditional name="predefined_cond">
+            <param name="predefined_sel" type="select" label="Use a predfinced OMERO server">
+                <option value="yes">Yes</option>
+                <option value="no">No</option>
+            </param>
+            <when value="yes">
+                <param name="omero_host" type="select" label="OMERO server">
+                    <options from_data_table="omero_servers"/>
```

Comment:
Keep in mind that users should eventually be able to bring their own data table entries, I'm not sure if this is going to be a problem. I think it might be better if this was a setting in extra user preferences ?

Reply:
> I think it might be better if this was a setting in extra user preferences ?

But we can not create options from this, or? 

> Keep in mind that users should eventually be able to bring their own data table entries, I'm not sure if this is going to be a problem.

Via tool data bundles? Have to admit that I still have only a rough intuition of this idea. If a user can choose the same ID for a data table entry credentials might be send to the wrong server ... sounds like an attack path. Otherwise users won't be able to use this because only the admin can define user preferences.

Reply:
> But we can not create options from this, or?

You can only have one setting value selected in the user preferences, but IIRC you can define a select in the user preferences. Getting access to user preference selects from the tool form would make sense to me.

I don't think server connections are really what we should be putting in tool data files. I'm not saying you can't do it, but I guess you asked me to review this to see if this would work as a general mechanism. If I was working on this and I'd implement another option source to handle this scenario.

Reply:
> You can only have one setting value selected in the user preferences, but IIRC you can define a select in the user preferences. 

Ah, I see - there is an example. There is an example. Should work such that admins can setup multiple OMERO servers. Each user can then use one of those .. but different users can use different servers. Might be sufficient.

> Getting access to user preference selects from the tool form would make sense to me.

Would then be sufficient in the command which already works, i.e. the user only chooses in the user preferences. And the tool then just uses this.